### ISSUE TITLE
`helm list`error maximum size exceeded

### ISSUE THREAD

**Issue by jascott1** [helm/helm#3322]
## Problem:
After installing kolla-kubernetes (a collection of helm charts resulting in many releases) and attempting to do a `helm list` from a remote client I receive the following error:

```
Error: grpc: received message larger than max (6645791 vs. 4194304)
```
This problem does not manifest when performing `helm list` on the server.
Currently the count of releases is 16 although I presume it also has to do with release size. Each release is composed of multiple sub-charts.

## Version:
```
Client: &version.Version{SemVer:"v2.7.2", GitCommit:"8478fb4fc723885b155c924d1c8c410b7a9444e6", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.7.2", GitCommit:"8478fb4fc723885b155c924d1c8c410b7a9444e6", GitTreeState:"clean"}
```

## Solutions:
Should I look into "chunking" the results as to not exceed the maximum or does someone have a better of idea of how to resolve this issue?

**Comment by technosophos** [comment]
We might need to use the streaming API. @fibonacci1729 do you know?

**Comment by jrthrawny** [comment]
I too am experiencing this issue

`Client: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}`

**Comment by wernight** [comment]
Same as @jrthrawny but noticed that I can successfully do:

    $ helm list --deployed

**Comment by rhefner1** [comment]
I'm stuck on this issue too but even the `--deployed` flag doesn't work:

```
[rhefner@platypus: ~]$ helm ls
Error: grpc: received message larger than max (6576945 vs. 4194304)
[rhefner@platypus: ~]$ helm ls --deployed
Error: grpc: received message larger than max (5311306 vs. 4194304)
```

I just upgrade to 2.8.0:
```
[rhefner@platypus: ~]$ helm version
Client: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}
```

**Comment by jrthrawny** [comment]
So there is a work around for this.

It seems that there are two issues that can cause it.

1. Binaries and other files from git are getting processed by helm, this can be fixed with [.helmignore](https://github.com/kubernetes/helm/blob/master/pkg/repo/repotest/testdata/examplechart/.helmignore)

2. `helm list` by my initial investigation will process all releases, failed or really old. We deployed tiller without a history limit and have several large charts with subcharts. The amount of revisions we had deployed were 30+. So to resolve this, use the sort, pagination and flags of the command:

`helm list --deployed -m 10`

There is no way to change history limit to my knowledge after Tiller has been deployed other than to completely uninstall it and reinstall. We did not want to do this, so if you go into the namespace that Tiller is deployed in(kube-system by default), you will find that Tiller keeps configmaps for every release and version of that release. We manually cleaned out the old revisions, up to just having the last 3. This solved that issue.

Perhaps this can be a feature request. The ability to purge revisions past a certain point.

**Comment by bacongobbler** [comment]
> There is no way to change history limit to my knowledge after Tiller has been deployed other than to completely uninstall it and reinstall.

You can upgrade helm in-place via `helm init --upgrade --history-max=10 ...`.

**Comment by jascott1** [comment]
In my case each release only has one revision (but there are many releases) so max history is not applicable.

**Comment by groundnuty** [comment]
Same problem. 
Fresh k8s:
~~~
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T10:09:24Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.2", GitCommit:"5fa2db2bd46ac79e5e00a4e6ed24191080aa463b", GitTreeState:"clean", BuildDate:"2018-01-18T09:42:01Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}
~~~
Fresh helm 
~~~
Client: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.0", GitCommit:"14af25f1de6832228539259b821949d20069a222", GitTreeState:"clean"}
~~~

Anything > 7 here:
~~~
helm list --deployed -m 7
~~~

helm list --deployed -m 7
~~~
     next: test
NAME                    REVISION        UPDATED                         STATUS          CHART                           NAMESPACE
develop                 1               Tue Feb  6 18:22:40 2018        DEPLOYED        cross-support-job-3p-0.2.6      develop
efk                     1               Mon Feb  5 18:26:12 2018        DEPLOYED        efk-0.1.0                       logs
fix-rc9-transfers       1               Sat Feb  3 13:57:49 2018        DEPLOYED        cross-support-job-3p-0.2.5      fix-rc9-transfers
go-carbon               1               Mon Feb  5 18:39:54 2018        DEPLOYED        go-carbon-0.1.0                 mon
grafana                 1               Mon Feb  5 18:39:54 2018        DEPLOYED        grafana-0.5.4                   mon
openvpn                 1               Fri Feb  2 06:16:19 2018        DEPLOYED        openvpn-2.0.3                   vpn
release17060-rc9        1               Mon Feb  5 13:58:21 2018        DEPLOYED        cross-support-job-3p-0.2.6      release17060-rc9
~~~

Not sure what the first line `next: test` is?

anything greater then 7 returns, `helm list --deployed -m 8` 
~~~
Error: grpc: received message larger than max (4687514 vs. 4194304)
~~~

**Comment by andor44** [comment]
Go-grpc got updated between 2.7-2.8 (check the release notes for the PR). It looks like they try to set max gRPC message size [here](https://github.com/kubernetes/helm/blob/838d7808946b554865e0fc897e7db8e8dfd63bde/pkg/tiller/server.go#L34-L39) but this seems to be no longer respected by go-grpc?

**Comment by mattfarina** [comment]
I think I found the problem...

When we set our options the code looks like:

https://github.com/kubernetes/helm/blob/838d7808946b554865e0fc897e7db8e8dfd63bde/pkg/tiller/server.go#L37-L43

This worked fine with grpc prior to update to 1.7.x where there was a change. The `MaxMsgSize` was broken into `MaxRecvMsgSize` and `MaxSendMsgSize`. The old function `MaxMsgSize` only updated the `MaxRecvMsgSize`. In helm the `MaxSendMsgSize` is now at 4mb.

**Comment by mattfarina** [comment]
Can y'all test the fix at #3469

**Comment by jascott1** [comment]
@mattfarina Good catch! It still doesn't fix my problem because I am actually exceeding the limit. Maybe we should look into streaming for helm 3.x?

**Comment by ccampo133** [comment]
Can confirm that we're using helm 2.8.1 and seeing this issue:

```$ helm list
Error: grpc: received message larger than max (4223965 vs. 4194304)```

**Comment by jascott1** [comment]
So should we try to stream the results or do we do the math for pagination and have the client make multiple calls, buffer those results and then display them?

**Comment by binoculars** [comment]
Same issue for me with 2.8.1:

```
Error: grpc: received message larger than max (9551184 vs. 4194304)
```

**Comment by bacongobbler** [comment]
Can someone confirm if this is an issue in prior versions, and if it isn't can they try to produce a reference example to test against?

**Comment by binoculars** [comment]
When I reset it back to v2.7.2 it was fine again.

**Comment by andor44** [comment]
It is still an issue on 2.8.1 client and server. I don't think any specific reference example is necessary other than "having a lot of releases" in Tiller.

**Comment by zhangtbj** [comment]
I see there i a PR about it: https://github.com/kubernetes/helm/pull/3514/files
`maxMsgSize use 20MB as the default message size limit.`
If the helm list is larger than 20MB, will it report the same error? @balboah 
And when will this PR be merged? Thanks!

**Comment by balboah** [comment]
@zhangtbj yes there is still a hard coded cap of 20MB instead of 4.
It will be merged when a maintainer approves it :)

**Comment by bacongobbler** [comment]
@balboah's PR was fixed, but OP's issue is still present since there's a 20MB cap on data that can be sent to and from helm and tiller in a single request.

There are ways to work around this by paging through the results as `helm help list` shows:

> By default, up to 256 items may be returned. To limit this, use the '--max' flag.
Setting '--max' to 0 will not return all results. Rather, it will return the
server's default, which may be much higher than 256. Pairing the '--max'
flag with the '--offset' flag allows you to page through results.

Perhaps a feature flag someone could introduce to `helm list` would be way to automate paging through the responses. Therefore, in @jascott1 and @zhangtbj's case it could be something like

```
$ helm list --paginate --max 128
```

Thoughts?

**Comment by bacongobbler** [comment]
re-labelling as a feature, since the bug was something else that got bunched up into this issue. :)

**Comment by pashau** [comment]
`helm install` fails with `Error: grpc: received message larger than max (5406598 vs. 4194304)` because Chart.tzg files from dependencies in my local helm chart repository are >4mb. 

```
Client: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}
```

**Comment by bacongobbler** [comment]
@pashau see #3514; you can test using the canary clients :) otherwise we should be able to cut a v2.8.2 release some time by end of week

**Comment by rofc** [comment]
helm v2.8.2 is working fine for me :)

**Comment by makhov** [comment]
We still have the issue. Created #3703 for fixing it.

**Comment by ultimateboy** [comment]
See https://github.com/kubernetes/helm/pull/3782 as well.

**Comment by purduemike** [comment]
This worked for me.

https://blog.shazam.com/rewriting-history-with-helm-b66e72958008

**Comment by bmarshall13** [comment]
I'm still seeing this with 2.10.0.

I was using 2.8.2 and upgraded some releases which triggered this bug. I upgraded thinking I would pick up some of the fixes mentioned in this issue, but after the fact I think 2.8.2 already had them.

```
$ helm version
Client: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.10.0", GitCommit:"9ad53aac42165a5fadc6c87be0dea6b115f93090", GitTreeState:"clean"}
```

```
$ helm ls
Error: trying to send message larger than max (24387804 vs. 20971520)
```

It looks like #3782 is trying to partition,  and is using a reasonable capacity. Tiller output for the above `helm ls`:
```
tiller-deploy-575fbb866c-n8xkv tiller [storage] 2018/09/07 15:51:02 listing all releases with filter
tiller-deploy-575fbb866c-n8xkv tiller [tiller] 2018/09/07 15:51:03 partitioned at 20337489 with 36 releases (cap=20971516)
tiller-deploy-575fbb866c-n8xkv tiller [tiller] 2018/09/07 15:51:03 partitioned at 20833363 with 31 releases (cap=20971516)
```

I can page through all the releases with `--max 10` and `--offset`, so I don't think there is an issue with the individual releases.

With --max 37, tiller still reports partitioning at 36, helm ls only prints 33 releases, and of those I get one duplicate.

**Comment by bacongobbler** [comment]
fixed in #4620. Thanks @bmarshall13!

**Comment by mr-yaky** [comment]
Unfortunately got the same error even with v2.11.0:
```
$ helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.1", GitCommit:"d4ab47518836c750f9949b9e0d387f20fb92260b", GitTreeState:"archive", BuildDate:"2018-04-26T09:29:05Z", GoVersion:"go1.10.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.0", GitCommit:"0ed33881dc4355495f623c6f22e7dd0b7632b7c0", GitTreeState:"clean", BuildDate:"2018-09-27T16:55:41Z", GoVersion:"go1.10.4", Compiler:"gc", Platform:"linux/amd64"}


helm install --name=rocketchat .
Error: grpc: received message larger than max (79313648 vs. 20971520)
```
